Autopilot/Src/IMU.cpp

```diff
@@ -0,0 +1,217 @@
+      
+      // Creates SPI Object
+      spi_port = new SPIPort(hspi_1);
+      GPIOPin pin(GPIO_PORT_A, 1, GPIO_OUTPUT, GPIO_STATE_LOW, GPIO_RES_NONE, GPIO_FREQ_HIGH, 0);                             //GET VALUES LATER 
+      imuSlaveIdentifier = spi_port->add_slave(pin); //Adds slave pin to the spi_port class object
+
+      uint8_t *setup = 0;
+
+      //Full reset of chip 
+      *setup = 0x80;
+      *imuBuffer = REG_PWR_MGMT_1 | 0x00; //0x80 means we are reading; 0x00 means we are writing
+      uint8_t writeData[2] = {*imuBuffer, *setup}; // writeData[0] -> register address; writeData[1] -> write instructions
+      write_data(writeData);
+
+      //Verify chip is working properly
+      *imuBuffer = REG_WHO_AM_I | 0x80;
+      uint8_t dataFromIMUSensor[2] = {0x00, 0x00}; // dataFromSensor[0] -> dummy byte (contains SPI address); dataFromSensor[1] -> data byte
+      spi_port->set_slave(imuSlaveIdentifier);
+      sensorSuccess = spi_port->exchange_data(imuBuffer, dataFromIMUSensor, 1);
+
+      if(sensorSuccess != STATUS_CODE_OK) {
+         //Debug code here
+      }
+
+      //Place accelerometer and gyroscopoe on standby
+      *setup = 0x3F;
+      *imuBuffer = REG_PWR_MGMT_2 | 0x00;
+      writeData[0] = *imuBuffer; 
+      writeData[1] = *setup; 
+      write_data(writeData); 
+
+      //Disable FIFO
+      *setup = 0x00;
+      *imuBuffer = REG_USER_CTRL | 0x00;
+      writeData[0] = *imuBuffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //Disable I2C
+      *setup = 0x40;
+      *imuBuffer = REG_I2C_IF | 0x00;
+      writeData[0] = *imuBuffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //Set up gyroscope
+      *setup = 0x10; //Setting to 500 dps max for gyroscope; setting to 4G max for accelerometer
+      *imuBuffer = REG_GYRO_CONFIG | 0x00;
+      writeData[0] = *imuBuffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //set up accelerometer
+      *imuBuffer = REG_ACCEL_CONFIG | 0x00; 
+      writeData[0] = *imuBuffer;
+      write_data(writeData); 
+
+      *setup = 0x08;
+      *imuBuffer = REG_ACCEL_CONFIG_2 | 0x00;
+      writeData[0] = *imuBuffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+  
+      //Turn on gyroscope and accelerometer
+      *setup = 0x00;
+      *imuBuffer = REG_PWR_MGMT_2 | 0x00;
+      writeData[0] = *imuBuffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+   }
+
+   accelConversionFactor = ACCEL_SENSITIVITY_4G;
+   gyroConversionFactor = GRYO_SENSITIVITY_500;
+   tempConversionFactor = TEMPERATURE_SENSITIVITY;
+}
+
+void ICM20602::write_data(uint8_t *writeData) { 
+   uint8_t *imuPlaceholderByte = 0x00;
+   spi_port->set_slave(imuSlaveIdentifier);
+   spi_port->exchange_data(writeData, imuPlaceholderByte, 2);                                                                    //WRITEDATA ARRAY HAS TWO ELEMENTS, THUS THE imuBuffer SIZE IS TWO HERE!!!!
```

Comment:
Is the buffer size (the integer 2) the correct value? I am passing in writeData, which has **two** uint8_t elements, as the tx_data parameter.

Reply:
THE imuPlaceholderByte should be a 2 element uint8_t array. 

The length parameter is correct